File: go/braidref/pulse.go
package braidref

import (
	"encoding/binary"
	"encoding/hex"
	"fmt"
	"math/big"

	"github.com/mindchain/noosphere/go/lumenref"
)

// Pulse v1 retarget per protocol/spec/pulse-exp2-v1.md. The exp2_q64 table
// is recomputed at init from the spec's §2 generator law with exact integer
// arithmetic and pinned to the frozen BLAKE3 hash; the §3 evaluation and
// rounding-order law is implemented verbatim (same table, same bit order,
// same per-step floors, same short circuits).

const (
	pulseTargetSpacingSeconds = 6
	pulseHalfLifeSeconds      = 3600
	// exp2TableHashHex is the frozen BLAKE3-256 over the 64 entries
	// concatenated as u128 little-endian (pulse-exp2-v1.md §2).
	exp2TableHashHex = "15d783a23bcf9d9e20d1133bbc247a5c94a876705aed5281e73246f67c883999"
)

// exp2Table holds EXP2_Q64_TABLE_V1[k] = floor(2^(2^-(k+1)) * 2^64) for
// k = 0..63; every entry is a 65-bit integer in [2^64, 2^65).
var exp2Table [64]*big.Int

// tMax is 2^256 - 1.
var tMax = new(big.Int).Sub(new(big.Int).Lsh(big.NewInt(1), 256), big.NewInt(1))

func init() {
	// Generator law (pulse-exp2-v1.md §2): work at P = 256 fractional
	// bits; s_1 = isqrt(2^(2P+1)); s_{j+1} = isqrt(s_j << P);
	// entry_{j-1} = s_j >> (P-64), with the truncation-unambiguity proof
	// (s >> (P-64)) == ((s+4) >> (P-64)) asserted at every step.
	const P = 256
	s := new(big.Int).Sqrt(new(big.Int).Lsh(big.NewInt(1), 2*P+1))
	four := big.NewInt(4)
	for j := 1; j <= 64; j++ {
		if j > 1 {
			s.Sqrt(new(big.Int).Lsh(s, P))
		}
		lo := new(big.Int).Rsh(s, P-64)
		hi := new(big.Int).Rsh(new(big.Int).Add(s, four), P-64)
		if lo.Cmp(hi) != 0 {
			panic("exp2 table: ambiguous truncation")
		}
		exp2Table[j-1] = lo
	}
	// Fixed-value checks (§2): entry_0 == isqrt(2^129), entry_63 == 2^64,
	// all entries in [2^64, 2^65) and strictly decreasing.
	one64 := new(big.Int).Lsh(big.NewInt(1), 64)
	one65 := new(big.Int).Lsh(big.NewInt(1), 65)
	if exp2Table[0].Cmp(new(big.Int).Sqrt(new(big.Int).Lsh(big.NewInt(1), 129))) != 0 {
		panic("exp2 table: entry 0")
	}
	if exp2Table[63].Cmp(one64) != 0 {
		panic("exp2 table: entry 63")
	}
	for k := range 64 {
		if exp2Table[k].Cmp(one64) < 0 || exp2Table[k].Cmp(one65) >= 0 {
			panic(fmt.Sprintf("exp2 table: entry %d out of range", k))
		}
		if k > 0 && exp2Table[k].Cmp(exp2Table[k-1]) >= 0 {
			panic(fmt.Sprintf("exp2 table: entry %d not decreasing", k))
		}
	}
	// Frozen table hash over u128-LE concatenation.
	buf := make([]byte, 0, 1024)
	var ent [16]byte
	for k := range 64 {
		lo := new(big.Int).And(exp2Table[k], new(big.Int).SetUint64(^uint64(0))).Uint64()
		hi := new(big.Int).Rsh(exp2Table[k], 64).Uint64()
		binary.LittleEndian.PutUint64(ent[:8], lo)
		binary.LittleEndian.PutUint64(ent[8:], hi)
		buf = append(buf, ent[:]...)
	}
	got := lumenref.DomainHash("", buf)
	want, _ := hex.DecodeString(exp2TableHashHex)
	if !equal32(got, want) {
		panic("exp2 table: hash mismatch against frozen constant")
	}
}

func equal32(a [32]byte, b []byte) bool {
	if len(b) != 32 {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

// PulseTargetV1 evaluates T_h = clamp(T_min, T_max,
// floor(T_a * 2^((t - t_a - 6*(h - h_a)) / 3600))) under the frozen §3
// rounding-order law.
//
// anchorTarget is T_a (contract: 1 <= T_a <= T_max); t is the parent
// block's median-time-past in integer seconds; (hA, tA) are the anchor
// height and median-time-past. Contract violations (h <= h_a, T_a out of
// range) are caller errors and panic rather than yielding a consensus
// verdict.
func PulseTargetV1(anchorTarget *big.Int, t, tA int64, h, hA uint64) *big.Int {
	if h <= hA {
		panic("pulse: h must exceed h_a")
	}
	if anchorTarget.Sign() < 1 || anchorTarget.Cmp(tMax) > 0 {
		panic("pulse: T_a out of range")
	}
	// 1. Exponent numerator, exact signed integer seconds.
	n := new(big.Int).SetInt64(t - tA)
	n.Sub(n, new(big.Int).Mul(big.NewInt(pulseTargetSpacingSeconds), new(big.Int).SetUint64(h-hA)))
	// 2. Euclidean division by 3600; f = floor(r * 2^64 / 3600).
	q, r := new(big.Int).DivMod(n, big.NewInt(pulseHalfLifeSeconds), new(big.Int))
	// big.Int.DivMod is Euclidean: 0 <= r < 3600.
	f := new(big.Int).Lsh(r, 64)
	f.Div(f, big.NewInt(pulseHalfLifeSeconds))
	// 3. Short circuits (exact equivalences, mandatory).
	if q.Cmp(big.NewInt(256)) >= 0 {
		return new(big.Int).Set(tMax)
	}
	if q.Cmp(big.NewInt(-257)) <= 0 {
		return big.NewInt(1)
	}
	// 4. Fractional walk, most-significant fractional bit first, flooring
	// at every step.
	acc := new(big.Int).Set(anchorTarget)
	for k := range 64 {
		if f.Bit(63-k) == 1 {
			acc.Mul(acc, exp2Table[k])
			acc.Rsh(acc, 64)
		}
	}
	// 5. Integer shift after the fractional walk.
	qi := q.Int64() // |q| <= 256 here
	if qi >= 0 {
		acc.Lsh(acc, uint(qi))
	} else {
		acc.Rsh(acc, uint(-qi))
	}
	// 6. Clamp into [1, 2^256 - 1].
	if acc.Sign() < 1 {
		return big.NewInt(1)
	}
	if acc.Cmp(tMax) > 0 {
		return new(big.Int).Set(tMax)
	}
	return acc
}

// U256FromLE converts a 32-byte little-endian unsigned integer.
func U256FromLE(b [32]byte) *big.Int {
	var be [32]byte
	for i := range b {
		be[31-i] = b[i]
	}
	return new(big.Int).SetBytes(be[:])
}

// U256ToLE converts to the canonical 32-byte little-endian form; v must be
// in [0, 2^256).
func U256ToLE(v *big.Int) [32]byte {
	var le [32]byte
	be := v.Bytes()
	for i, x := range be {
		le[len(be)-1-i] = x
	}
	return le
}

File: crates/noos-ground/src/pulse.rs
//! Pulse v1 — the deterministic ASERT retarget for Ground targets.
//!
//! Interface law (ch01 §4.2/§4.3; plan §6.2):
//!
//! ```text
//! T_h = clamp(T_min, T_max, floor(T_a * 2^((t - t_a - Δ(h - h_a)) / τ)))
//! ```
//!
//! with target spacing `Δ = 6 s`, half-life `τ = 3600 s`, `T_min = 1`,
//! `T_max = 2^256 - 1`; `t` is the **parent** median-time-past in seconds
//! and the anchor `(h_a, t_a, T_a)` is the most recent finalized checkpoint
//! on the branch (its already-validated height, median-time-past, and
//! target — never local arrival time; the function is pure).
//!
//! ## Exact evaluation law (frozen; Go must reproduce bit-for-bit)
//!
//! The full rounding-order law is specified in
//! `protocol/spec/pulse-exp2-v1.md`; summary:
//!
//! 1. `n = t - t_a - 6*(h - h_a)` in signed exact integer seconds.
//! 2. Q64.64 exponent rounded toward negative infinity:
//!    `q = n div_euclid 3600`, `r = n rem_euclid 3600` (`0 <= r < 3600`),
//!    `f = floor(r * 2^64 / 3600)`; the exponent is `q + f * 2^-64`.
//! 3. Short circuits (exactly equivalent to the long computation):
//!    `q >= 256` yields `T_max`; `q <= -257` yields `T_min`.
//! 4. Accumulator `acc = T_a` as a 512-bit integer. For `k = 0..63` in
//!    ascending order (most-significant fractional bit first), if bit
//!    `63 - k` of `f` is set: `acc = floor(acc * EXP2_Q64_TABLE_V1[k] / 2^64)`.
//! 5. Integer part after the fractional walk: `q >= 0` shifts left by `q`,
//!    `q < 0` shifts right by `-q` (floor).
//! 6. Clamp into `[T_min, T_max]`.

use crate::exp2_table::EXP2_Q64_TABLE_V1;
use crate::u256::{U256, U512};
use core::fmt;

/// Pulse v1 target spacing `Δ` in seconds (constants-v1.toml `[pulse]`).
pub const TARGET_SPACING_SECONDS: u64 = 6;
/// Pulse v1 half-life `τ` in seconds.
pub const HALF_LIFE_SECONDS: u64 = 3600;
/// Minimum target `T_min = 1`.
pub const T_MIN: U256 = U256::ONE;
/// Maximum target `T_max = 2^256 - 1`.
pub const T_MAX: U256 = U256::MAX;

/// The retarget anchor: the most recent finalized checkpoint on the branch.
///
/// All three fields come from that checkpoint's already-validated header
/// chain (ch01 §4.3: rolling the anchor on finality "does not consult local
/// arrival time").
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct PulseAnchor {
    /// Checkpoint height `h_a`.
    pub height: u64,
    /// Checkpoint median-time-past `t_a`, seconds.
    pub median_time_past_s: u64,
    /// Checkpoint Ground target `T_a`; must lie in `[T_min, T_max]`.
    pub target: U256,
}

/// Pulse evaluation errors. All are caller-contract violations; a valid
/// finalized checkpoint and a descendant height never produce one.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum PulseError {
    /// `child_height <= anchor.height`: the child must descend from the
    /// anchor checkpoint.
    ChildNotAfterAnchor {
        /// Anchor height `h_a`.
        anchor_height: u64,
        /// Offending child height `h`.
        child_height: u64,
    },
    /// The anchor target is zero (below `T_min`); no valid checkpoint
    /// carries it.
    AnchorTargetZero,
    /// Internal width overflow. Unreachable under the documented bounds;
    /// kept as a checked error rather than a panic path.
    Overflow,
}

impl fmt::Display for PulseError {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        match self {
            Self::ChildNotAfterAnchor {
                anchor_height,
                child_height,
            } => write!(
                f,
                "child height {child_height} does not descend from anchor height {anchor_height}"
            ),
            Self::AnchorTargetZero => f.write_str("anchor target is zero (below T_min)"),
            Self::Overflow => f.write_str("internal width overflow (bounds violated)"),
        }
    }
}

/// Signed Q64.64 exponent `(t - t_a - Δ(h - h_a)) / τ`, rounded toward
/// negative infinity, split into integer part `q` and fractional part
/// `f ∈ [0, 2^64)` such that the exponent equals `q + f * 2^-64`.
fn exponent_q64(
    anchor: &PulseAnchor,
    parent_median_time_past_s: u64,
    child_height: u64,
) -> Result<(i128, u64), PulseError> {
    let dh = child_height
        .checked_sub(anchor.height)
        .filter(|&dh| dh > 0)
        .ok_or(PulseError::ChildNotAfterAnchor {
            anchor_height: anchor.height,
            child_height,
        })?;
    // All terms fit i128 comfortably: u64 values and 6 * 2^64 < 2^67.
    let drift = i128::from(parent_median_time_past_s)
        .checked_sub(i128::from(anchor.median_time_past_s))
        .and_then(|d| {
            i128::from(TARGET_SPACING_SECONDS)
                .checked_mul(i128::from(dh))
                .and_then(|spacing| d.checked_sub(spacing))
        })
        .ok_or(PulseError::Overflow)?;
    let tau = i128::from(HALF_LIFE_SECONDS);
    // floor(n * 2^64 / 3600) == q * 2^64 + floor(r * 2^64 / 3600) with
    // n = 3600q + r, 0 <= r < 3600: Euclidean division rounds q toward
    // negative infinity and the fractional term is exact in u128
    // (r * 2^64 < 2^76).
    let q = drift.div_euclid(tau);
    let r = drift.rem_euclid(tau) as u128;
    // Divisor is the nonzero constant 3600; the fallback is unreachable.
    let f = (r << 64)
        .checked_div(u128::from(HALF_LIFE_SECONDS))
        .unwrap_or(0) as u64;
    Ok((q, f))
}

/// Deterministic Pulse v1 output: the Ground target every child of this
/// parent must carry (ch01 §4.2 rule 5).
///
/// Pure function of `(anchor, parent median-time-past, child height)`;
/// see the module docs for the frozen evaluation and rounding-order law.
pub fn pulse_target_v1(
    anchor: &PulseAnchor,
    parent_median_time_past_s: u64,
    child_height: u64,
) -> Result<U256, PulseError> {
    if anchor.target.is_zero() {
        return Err(PulseError::AnchorTargetZero);
    }
    let (q, f) = exponent_q64(anchor, parent_median_time_past_s, child_height)?;

    // Short circuits — exact equivalences, not approximations:
    // * every table entry is >= 2^64, so each fractional step is
    //   non-decreasing and acc >= T_a >= 1 after the walk; `acc << q`
    //   with q >= 256 is >= 2^256 > T_max, hence clamps to T_max.
    // * each fractional factor is < 2^65, so acc < 2*T_a <= 2^257 after
    //   the walk; `acc >> 257` (or more) is 0, hence clamps to T_min.
    if q >= 256 {
        return Ok(T_MAX);
    }
    if q <= -257 {
        return Ok(T_MIN);
    }

    // Fractional walk, most-significant exponent bit first.
    let mut acc = U512::from_u256(&anchor.target);
    for (k, &entry) in EXP2_Q64_TABLE_V1.iter().enumerate() {
        let bit = 63_u32.wrapping_sub(k as u32);
        if (f >> bit) & 1 == 1 {
            acc = acc.mul_shr64(entry).ok_or(PulseError::Overflow)?;
        }
    }

    // Integer shift, then clamp.
    let shifted = if q >= 0 {
        // q in [0, 255] and acc < 2^257: fits 512 bits by construction,
        // but stay checked.
        acc.checked_shl(q as u32).ok_or(PulseError::Overflow)?
    } else {
        // q in [-256, -1]: right shift floors.
        acc.shr(q.unsigned_abs() as u32)
    };
    if shifted.is_zero() {
        return Ok(T_MIN);
    }
    Ok(shifted.try_into_u256().unwrap_or(T_MAX))
}

#[cfg(test)]
#[allow(
    clippy::unwrap_used,
    clippy::expect_used,
    clippy::arithmetic_side_effects
)]
mod tests {
    use super::*;

    fn anchor(target: U256) -> PulseAnchor {
        PulseAnchor {
            height: 1024,
            median_time_past_s: 100_000_000,
            target,
        }
    }

    /// parent MTP that makes the exponent exactly `seconds / 3600`.
    fn mtp_for(anchor: &PulseAnchor, child_height: u64, seconds: i64) -> u64 {
        let on_schedule =
            anchor.median_time_past_s + TARGET_SPACING_SECONDS * (child_height - anchor.height);
        on_schedule.checked_add_signed(seconds).unwrap()
    }

    #[test]
    fn exponent_zero_is_identity() {
        let a = anchor(U256::from_u128(0xdead_beef_cafe_f00d_1234));
        let h = a.height + 100;
        let t = mtp_for(&a, h, 0);
        assert_eq!(pulse_target_v1(&a, t, h).unwrap(), a.target);
    }

    #[test]
    fn exponent_plus_one_doubles_exactly() {
        let a = anchor(U256::from_u128(12345678901234567890));
        let h = a.height + 1;
        let t = mtp_for(&a, h, 3600);
        assert_eq!(
            pulse_target_v1(&a, t, h).unwrap(),
            U256::from_u128(2 * 12345678901234567890)
        );
    }

    #[test]
    fn exponent_minus_one_halves_with_floor() {
        let odd = U256::from_u64(0xffff_ffff_ffff_fff1); // odd value
        let a = anchor(odd);
        let h = a.height + 1;
        let t = mtp_for(&a, h, -3600);
        assert_eq!(
            pulse_target_v1(&a, t, h).unwrap(),
            U256::from_u64(0xffff_ffff_ffff_fff1 / 2)
        );
    }

    #[test]
    fn clamps_at_t_max() {
        let a = anchor(U256::MAX);
        let h = a.height + 1;
        // +1 exponent on T_max must clamp, not wrap.
        let t = mtp_for(&a, h, 3600);
        assert_eq!(pulse_target_v1(&a, t, h).unwrap(), T_MAX);
        // Enormous positive exponent short-circuit.
        let t = mtp_for(&a, h, 3600 * 10_000);
        assert_eq!(pulse_target_v1(&a, t, h).unwrap(), T_MAX);
    }

    #[test]
    fn clamps_at_t_min() {
        let a = anchor(U256::ONE);
        let h = a.height + 1;
        let t = mtp_for(&a, h, -3600);
        assert_eq!(pulse_target_v1(&a, t, h).unwrap(), T_MIN);
        // Deep negative exponent short-circuit.
        let t = mtp_for(&a, h, -3600 * 10_000);
        assert_eq!(pulse_target_v1(&a, t, h).unwrap(), T_MIN);
        // Large target driven all the way down.
        let a = anchor(U256::MAX);
        let t = mtp_for(&a, h, -3600 * 300);
        assert_eq!(pulse_target_v1(&a, t, h).unwrap(), T_MIN);
    }

    #[test]
    fn short_circuit_boundaries_match_long_path() {
        // q = 255 (long path) vs q = 256 (short circuit): both T_max for a
        // large anchor target; and the q = -256 long path vs q = -257 short
        // circuit agree for T_a = 1.
        let a = anchor(U256::from_u64(1));
        let h = a.height + 1;
        let t = mtp_for(&a, h, 3600 * 256);
        assert_eq!(pulse_target_v1(&a, t, h).unwrap(), T_MAX);
        let t = mtp_for(&a, h, 3600 * 255);
        // 1 << 255 fits: exact power of two.
        let mut limbs = [0_u64; 4];
        limbs[3] = 1 << 63;
        assert_eq!(pulse_target_v1(&a, t, h).unwrap(), U256::from_limbs(limbs));
        let t = mtp_for(&a, h, -3600 * 256);
        assert_eq!(pulse_target_v1(&a, t, h).unwrap(), T_MIN);
        let t = mtp_for(&a, h, -3600 * 257);
        assert_eq!(pulse_target_v1(&a, t, h).unwrap(), T_MIN);
    }

    #[test]
    fn half_exponent_multiplies_by_sqrt2_entry() {
        // n = +1800 s => q = 0, f = 2^63 => acc = floor(T_a * table[0] / 2^64).
        let ta = 1_u128 << 64;
        let a = anchor(U256::from_u128(ta));
        let h = a.height + 1;
        let t = mtp_for(&a, h, 1800);
        assert_eq!(
            pulse_target_v1(&a, t, h).unwrap(),
            U256::from_u128(crate::exp2_table::EXP2_Q64_TABLE_V1[0])
        );
    }

    #[test]
    fn fractional_rounding_is_toward_negative_infinity() {
        // n = -1 s: exponent = -1/3600, i.e. q = -1, r = 3599.
        // Result must be strictly below T_a and >= floor(T_a * 2^-1):
        // exact value cross-checked against the Python oracle in the
        // conformance vectors; here we pin the floor direction.
        let ta = U256::from_u128((1 << 100) + 12345);
        let a = anchor(ta);
        let h = a.height + 1;
        let t = mtp_for(&a, h, -1);
        let out = pulse_target_v1(&a, t, h).unwrap();
        assert!(out < ta);
        // 2^-1/3600 > 0.9998; output must be above 0.999 * T_a:
        // floor(T_a * 999 / 1000) as a coarse lower bound.
        let coarse = U512::from_u256(&ta)
            .mul_shr64(((999_u128) << 64) / 1000)
            .unwrap()
            .try_into_u256()
            .unwrap();
        assert!(out > coarse);
    }

    #[test]
    fn determinism_and_arrival_time_independence() {
        // Pure function: repeated evaluation and permuted evaluation order
        // give identical bytes; there is no clock input at all.
        let a = anchor(U256::from_u128(0xabcdef0123456789));
        let h = a.height + 7;
        let t = mtp_for(&a, h, 12_345);
        let first = pulse_target_v1(&a, t, h).unwrap();
        for _ in 0..10 {
            assert_eq!(pulse_target_v1(&a, t, h).unwrap(), first);
        }
    }

    #[test]
    fn contract_violations_error() {
        let a = anchor(U256::from_u64(5));
        assert_eq!(
            pulse_target_v1(&a, 0, a.height),
            Err(PulseError::ChildNotAfterAnchor {
                anchor_height: a.height,
                child_height: a.height
            })
        );
        let z = anchor(U256::ZERO);
        assert_eq!(
            pulse_target_v1(&z, 0, z.height + 1),
            Err(PulseError::AnchorTargetZero)
        );
    }
}
